### PR TITLE
feat: the empty line in commands linter

### PR DESCRIPTION
**Pull request by adomani**
This linter flags empty lines within a command.

It allows empty lines within doc-strings, module-docs and a couple of other "sensible" places.

It also skips files that are likely to contain meta-code, since there the use of empty lines in definition is more widespread.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Tactic/Linter/EmptyLine.lean [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+namespace EmptyLine
+
+/-- The `SyntaxNodeKind`s where the `EmptyLine` linter is not active. -/
+abbrev AllowEmptyLines : Std.HashSet SyntaxNodeKind := Std.HashSet.emptyWithCapacity
+  |>.insert ``Parser.Command.docComment
+  |>.insert ``Parser.Command.moduleDoc
+  |>.insert ``Parser.Command.mutual
+  |>.insert `str
+
+/-- If a file contains one of these names as segments, we disable `EmptyLine` linter. -/
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/EmptyLine.lean:58]:
```suggestion
/-- If a file contains one of these names as segments, we disable `emptyLine` linter. -/

File: Mathlib/Tactic/Linter/EmptyLine.lean [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+@[inherit_doc Mathlib.Linter.linter.style.emptyLine]
+def emptyLineLinter : Linter where run := withSetOptionIn fun stx ↦ do
+  unless Linter.getLinterValue linter.style.emptyLine (← getOptions) do
+    return
+  if (← get).messages.hasErrors then
+    return
+  if !((← getMainModule).components.filter (SkippedFileSegments.contains)).isEmpty then
+    return
+  -- We ignore empty lines "after" the command finished
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/EmptyLine.lean:72]:
```suggestion
  -- We ignore empty lines "after" the command finished.

**Review by grunweg — COMMENTED** [review]
Two nits. The fixes so far LGTM.

**Comment by bryangingechen** [comment]
Could you (or someone else) open a PR to the style guide also so we can have some guidance for contributors in place before this is enforced?

**Comment by adomani** [comment]
@bryangingechen thank you for the suggestion!  [community#637](https://github.com/leanprover-community/leanprover-community.github.io/pull/637)

**Comment by leanprover-community-bot-assistant** [comment]
This pull request has conflicts, please merge `master` and resolve them.

File: Mathlib/Tactic/Linter/EmptyLine.lean [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+incomplete proofs.
+-/
+register_option linter.style.emptyLine : Bool := {
+  defValue := false
+  descr := "enable the emptyLine linter"
+}
+
+namespace EmptyLine
+
+/-- The `SyntaxNodeKind`s where the `EmptyLine` linter is not active. -/
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/EmptyLine.lean]: remove the full stop?

File: Mathlib/Tactic/Linter/EmptyLine.lean [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+abbrev AllowEmptyLines : Std.HashSet SyntaxNodeKind := Std.HashSet.emptyWithCapacity
+  |>.insert ``Parser.Command.docComment
+  |>.insert ``Parser.Command.moduleDoc
+  |>.insert ``Parser.Command.mutual
+  |>.insert `str
+
+/-- If a file contains one of these names as segments, we disable the `emptyLine` linter. -/
+abbrev SkippedFileSegments : Std.HashSet Name := Std.HashSet.emptyWithCapacity
+  |>.insert `Tactic
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/EmptyLine.lean:85]: Add a test for one of these?

File: Mathlib/Tactic/Linter/EmptyLine.lean [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+  |>.insert `Util
+  |>.insert `Meta
+
+@[inherit_doc Mathlib.Linter.linter.style.emptyLine]
+def emptyLineLinter : Linter where run := withSetOptionIn fun stx ↦ do
+  unless Linter.getLinterValue linter.style.emptyLine (← getOptions) do
+    return
+  -- The linter does not report anything on incomplete proofs.
+  if (← get).messages.reportedPlusUnreported.any (!·.severity matches .information) then
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/EmptyLine.lean:95]: This check is different from the analogous checks in the other syntax linters. Why so? Is this one equivalent, more thorough (so a PR to the other linters is in order), different (how)?

File: Mathlib/Tactic/Linter/EmptyLine.lean [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+      let trails := stx.filterMap fun s =>
+        if let some str := s.getTrailing?
+        then
+          -- Handle `where` and `where` fields.
+          if s.getAtomVal == "where" ||
+             s.isOfKind ``Parser.Term.structInstField ||
+             s.isOfKind ``Parser.Command.structSimpleBinder then
+            s.getTrailing?.map (·.getRange)
+          else
+          let strim := str.trim
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/EmptyLine.lean]: Indent these lines?

**Review by grunweg — COMMENTED** [review]
Thanks! The linter implementation looks good to me. One main question (and I haven't reviewed the tests yet). I pushed a few nits directly.

**Comment by grunweg** [comment]
The tests also looks nice and thorough. Probably, there may be edge cases that we'll discover, but lots of them are already tested - this looks safe to try.

File: Mathlib/Tactic/Linter/EmptyLine.lean (hunk 4 above)
Comment (grunweg) [reply Mathlib/Tactic/Linter/EmptyLine.lean:85]: This is possible, but would mean adding a new file whose path matches one of these segments. I don't think this is necessarily worth it, and am fine landing this without.

**Comment by grunweg** [comment]
So, apart from the one question I asked above, this looks ready for me.

File: Mathlib/Tactic/Linter/EmptyLine.lean [hunk 7]
```diff
@@ -100,8 +100,7 @@ def emptyLineLinter : Linter where run := withSetOptionIn fun stx ↦ do
   let allowed := stx.filter (AllowEmptyLines.contains ·.getKind)
   let allowedRanges := allowed.filterMap (·.getRange?)
   if let some str := stx.getSubstring? then
-    if let one::two::twos := str.toString.trimRight.splitOn "\n\n" then
-      let rest := two::twos
+    if let one::rest := str.toString.trimRight.splitOn "\n\n" then
```
Comment (adomani) [inline Mathlib/Tactic/Linter/EmptyLine.lean:104]: Actually, I think that the earlier version is what I intended: the older version was making sure that the list had at least 2 entries, whereas now `rest` could be empty.

File: Mathlib/Tactic/Linter/EmptyLine.lean (hunk 5 above)
Comment (adomani) [reply Mathlib/Tactic/Linter/EmptyLine.lean:95]: This linter will not report when a proof is noisy, not just when there are errors.  Thus, an incomplete proof with `sorry`s filled in will also silence this linter.

File: Mathlib/Tactic/Linter/EmptyLine.lean (hunk 7 above)
Comment (grunweg) [reply Mathlib/Tactic/Linter/EmptyLine.lean:104]: :bulb: :facepalm: Right! Feel free to revert (and, maybe, add a comment to help future Michael).

File: Mathlib/Tactic/Linter/EmptyLine.lean (hunk 5 above)
Comment (grunweg) [reply Mathlib/Tactic/Linter/EmptyLine.lean:95]: Ah. To rephrase: most syntax linters become silent when a proof has errors, but *do* report when there are (other) warnings, such as remaining sorries. *This linter* is silent as long as there is any warning. Correct?

Comment (adomani) [reply Mathlib/Tactic/Linter/EmptyLine.lean:95]: I expanded a comment above this check to mention `sorry` and `stop`.

File: Mathlib/Tactic/Linter/EmptyLine.lean (hunk 7 above)
Comment (adomani) [reply Mathlib/Tactic/Linter/EmptyLine.lean:104]: I liked the other changes, so I simply added a check that `rest` is non-empty: that should be sufficiently explicit, I hope!

Comment (adomani) [reply Mathlib/Tactic/Linter/EmptyLine.lean:104]: (Btw, I think that this is just a potential performance optimisation: the linter loops over the entries of `rest`, so will do nothing anyway if `rest` is empty, but it precomputes something before looping.  I am not sure if the pre-computations would be performed if `rest` is empty or not.  With the extra check, I am hoping that they will not be performed!)

File: Mathlib/Tactic/Linter/EmptyLine.lean (hunk 4 above)
Comment (adomani) [reply Mathlib/Tactic/Linter/EmptyLine.lean:85]: I think that this is not really worth it: there are files with `Tactic` in their name, so, in some sense, those files serve as tests already.

Besides, maybe in the future the linter should expand to look at "meta" files, maybe with some more refined heuristic than the current one.

**Comment by grunweg** [comment]
Thanks!
maintainer merge

**Comment by adomani** [comment]
This PR has been migrated to a fork-based workflow: https://github.com/leanprover-community/mathlib4/pull/25945
